File: firmware/components/calc_engine/calc_engine.cpp
#include "calc_engine.h"

#include <cmath>

double calc_evaluate_at(const expr_handle_t *expr, double x, angle_mode_t mode) {
    return expr_eval_at(expr, x, mode);
}

bool calc_find_root(const expr_handle_t *expr, double x_min, double x_max,
                    angle_mode_t mode, double guess_x, double *out_x) {
    if (!expr || !out_x) return false;
    const int steps = 200;
    const double dx = (x_max - x_min) / steps;
    double closest = NAN;
    double closest_dist = INFINITY;

    for (int i = 0; i < steps; ++i) {
        const double x1 = x_min + i * dx;
        const double x2 = x1 + dx;
        const double y1 = calc_evaluate_at(expr, x1, mode);
        const double y2 = calc_evaluate_at(expr, x2, mode);
        if (!std::isfinite(y1) || !std::isfinite(y2)) continue;

        if (std::fabs(y1) < 1e-12) {
            const double dist = std::fabs(x1 - guess_x);
            if (dist < closest_dist) {
                closest = x1;
                closest_dist = dist;
            }
        }

        if (y1 * y2 < 0.0) {
            double a = x1, b = x2;
            double root = NAN;
            for (int iter = 0; iter < 100; ++iter) {
                const double mid = (a + b) / 2.0;
                const double yMid = calc_evaluate_at(expr, mid, mode);
                if (std::fabs(yMid) < 1e-12 || (b - a) / 2.0 < 1e-12) {
                    root = mid;
                    break;
                }
                const double yA = calc_evaluate_at(expr, a, mode);
                if (yA * yMid < 0.0) b = mid;
                else a = mid;
            }
            if (std::isfinite(root)) {
                const double dist = std::fabs(root - guess_x);
                if (dist < closest_dist) {
                    closest = root;
                    closest_dist = dist;
                }
            }
        }
    }

    if (!std::isfinite(closest)) return false;
    *out_x = closest;
    return true;
}

bool calc_find_extremum(const expr_handle_t *expr, double x_min, double x_max,
                        angle_mode_t mode, bool find_max, double guess_x,
                        calc_point_t *out) {
    if (!expr || !out) return false;
    const int steps = 500;
    const double dx = (x_max - x_min) / steps;

    double cand_x[64];
    int cand_n = 0;

    double prev_y = NAN;
    double prev_slope = NAN;

    for (int i = 0; i <= steps; ++i) {
        const double x = x_min + i * dx;
        const double y = calc_evaluate_at(expr, x, mode);
        if (!std::isfinite(y)) {
            prev_y = NAN;
            prev_slope = NAN;
            continue;
        }
        if (std::isfinite(prev_y)) {
            const double slope = (y - prev_y) / dx;
            if (std::isfinite(prev_slope)) {
                if (prev_slope > 0 && slope < 0 && find_max && cand_n < 64) {
                    cand_x[cand_n++] = x - dx / 2.0;
                } else if (prev_slope < 0 && slope > 0 && !find_max && cand_n < 64) {
                    cand_x[cand_n++] = x - dx / 2.0;
                }
            }
            prev_slope = slope;
        }
        prev_y = y;
    }

    if (cand_n == 0) return false;

    int best = 0;
    double min_diff = std::fabs(cand_x[0] - guess_x);
    for (int i = 1; i < cand_n; ++i) {
        const double d = std::fabs(cand_x[i] - guess_x);
        if (d < min_diff) {
            min_diff = d;
            best = i;
        }
    }

    double x = cand_x[best];
    double step = dx / 2.0;
    for (int iter = 0; iter < 30; ++iter) {
        const double y = calc_evaluate_at(expr, x, mode);
        const double yL = calc_evaluate_at(expr, x - step, mode);
        const double yR = calc_evaluate_at(expr, x + step, mode);
        if (find_max) {
            if (std::isfinite(yL) && std::isfinite(yR)) {
                if (yL > y && yL > yR) x -= step;
                else if (yR > y && yR > yL) x += step;
                else step /= 2.0;
            } else step /= 2.0;
        } else {
            if (std::isfinite(yL) && std::isfinite(yR)) {
                if (yL < y && yL < yR) x -= step;
                else if (yR < y && yR < yL) x += step;
                else step /= 2.0;
            } else step /= 2.0;
        }
    }

    const double final_y = calc_evaluate_at(expr, x, mode);
    if (!std::isfinite(final_y)) return false;
    out->x = x;
    out->y = final_y;
    out->valid = true;
    return true;
}

double calc_derivative(const expr_handle_t *expr, double x, angle_mode_t mode) {
    const double h = 1e-5;
    const double y1 = calc_evaluate_at(expr, x - h, mode);
    const double y2 = calc_evaluate_at(expr, x + h, mode);
    if (!std::isfinite(y1) || !std::isfinite(y2)) return NAN;
    return (y2 - y1) / (2.0 * h);
}

double calc_integrate(const expr_handle_t *expr, double lower, double upper, angle_mode_t mode) {
    const int N = 1000;
    const double h = (upper - lower) / N;
    const double yStart = calc_evaluate_at(expr, lower, mode);
    const double yEnd = calc_evaluate_at(expr, upper, mode);
    if (!std::isfinite(yStart) || !std::isfinite(yEnd)) return NAN;
    double sum = 0.5 * (yStart + yEnd);
    for (int i = 1; i < N; ++i) {
        const double y = calc_evaluate_at(expr, lower + i * h, mode);
        if (!std::isfinite(y)) return NAN;
        sum += y;
    }
    return sum * h;
}

double calc_second_derivative(const expr_handle_t *expr, double x, angle_mode_t mode) {
    const double h = 1e-3;
    const double yP = calc_evaluate_at(expr, x + h, mode);
    const double yM = calc_evaluate_at(expr, x - h, mode);
    const double yV = calc_evaluate_at(expr, x, mode);
    if (!std::isfinite(yP) || !std::isfinite(yM) || !std::isfinite(yV)) return NAN;
    return (yP - 2.0 * yV + yM) / (h * h);
}

double calc_find_inflection(const expr_handle_t *expr, double x1, double x2, angle_mode_t mode) {
    double a = x1, b = x2;
    double lastMid = (a + b) / 2.0;
    for (int iter = 0; iter < 15; ++iter) {
        const double mid = (a + b) / 2.0;
        const double fMid = calc_second_derivative(expr, mid, mode);
        const double fA = calc_second_derivative(expr, a, mode);
        if (!std::isfinite(fMid) || !std::isfinite(fA)) return lastMid;
        if (std::fabs(fMid) < 1e-8) return mid;
        if (fA * fMid < 0.0) b = mid;
        else a = mid;
        lastMid = mid;
    }
    return lastMid;
}

File: CalcEngine.js
/**
 * CalcEngine.js
 * Numerical solvers and expression evaluation at X. DOM-free.
 * Expression backend is global `math` (math.js) via compiled exprs; ESP replaces with TinyExpr.
 */

class CalcEngine {
  /**
   * Coerce any complex/fraction/bigNumber to a real number.
   * Imaginary residuals below 1e-12 are treated as real.
   */
  static coerceToReal(val) {
    if (typeof val === 'number') {
      return val;
    }
    if (val && typeof val.toNumber === 'function') {
      return val.toNumber();
    }
    if (val && typeof val.im === 'number' && typeof val.re === 'number') {
      return Math.abs(val.im) < 1e-12 ? val.re : NaN;
    }
    return NaN;
  }

  /**
   * Formats a value, handling numbers and math.js complex objects.
   */
  static formatComplexValue(val, precisionMode) {
    if (typeof val === 'number') {
      return CalcEngine.formatReal(val, precisionMode);
    }

    // Check if it is a complex number from math.js
    if (val && typeof val.re === 'number' && typeof val.im === 'number') {
      const re = val.re;
      const im = val.im;

      // If imaginary part is practically zero, format as real
      if (Math.abs(im) < 1e-12) {
        return CalcEngine.formatReal(re, precisionMode);
      }

      // If real part is practically zero, format as imaginary
      if (Math.abs(re) < 1e-12) {
        return CalcEngine.formatImaginary(im, precisionMode);
      }

      // Both real and imaginary parts are present: a + bi or a - bi
      const reStr = CalcEngine.formatReal(re, precisionMode);
      const imStr = CalcEngine.formatImaginary(Math.abs(im), precisionMode);
      const sign = im > 0 ? ' + ' : ' - ';

      return `${reStr}${sign}${imStr}`;
    }

    if (val && typeof val.toString === 'function') {
      return val.toString();
    }
    return String(val);
  }

  static formatReal(val, precisionMode) {
    if (typeof val !== 'number' || isNaN(val)) return "NaN";
    if (!isFinite(val)) return val > 0 ? "Infinity" : "-Infinity";
    if (precisionMode === 'float') {
      return parseFloat(val.toFixed(6)).toString();
    }
    const prec = parseInt(precisionMode.slice(3)) || 2;
    return val.toFixed(prec);
  }

  static formatImaginary(im, precisionMode) {
    if (im === 1) return 'i';
    if (im === -1) return '-i';
    const formatted = CalcEngine.formatReal(im, precisionMode);
    return `${formatted}i`;
  }

  /**
   * Evaluates the equation at X, respecting Angle Mode (rad vs deg).
   * Always returns a coerced real number, or NaN.
   * @param {Object} compiled - Compiled math.js expression
   * @param {number} xVal - The X coordinate
   * @param {string} angleMode - 'rad' or 'deg'
   * @returns {number}
   */
  static evaluateAt(compiled, xVal, angleMode) {
    if (!compiled) return NaN;
    try {
      const scope = { x: xVal };
      if (angleMode === 'deg') {
        scope.sin = (val) => Math.sin(val * Math.PI / 180);
        scope.cos = (val) => Math.cos(val * Math.PI / 180);
        scope.tan = (val) => Math.tan(val * Math.PI / 180);
        scope.asin = (val) => Math.asin(val) * 180 / Math.PI;
        scope.acos = (val) => Math.acos(val) * 180 / Math.PI;
        scope.atan = (val) => Math.atan(val) * 180 / Math.PI;
      }
      const raw = compiled.evaluate(scope);
      return CalcEngine.coerceToReal(raw);
    } catch {
      return NaN;
    }
  }

  /**
   * Finds a root of f(x) in [xMin, xMax] using bisection/scanning.
   */
  static findRoot(compiled, xMin, xMax, angleMode, guessX = 0) {
    const steps = 200;
    const dx = (xMax - xMin) / steps;
    let closestRoot = null;
    let closestDist = Infinity;

    for (let i = 0; i < steps; i++) {
      const x1 = xMin + i * dx;
      const x2 = x1 + dx;

      let y1, y2;
      try {
        y1 = this.evaluateAt(compiled, x1, angleMode);
        y2 = this.evaluateAt(compiled, x2, angleMode);
      } catch {
        continue;
      }

      if (isNaN(y1) || isNaN(y2) || !isFinite(y1) || !isFinite(y2)) continue;

      if (Math.abs(y1) < 1e-12) {
        const dist = Math.abs(x1 - guessX);
        if (dist < closestDist) {
          closestRoot = x1;
          closestDist = dist;
        }
      }

      if (y1 * y2 < 0) {
        let a = x1;
        let b = x2;
        let root = null;
        for (let iter = 0; iter < 100; iter++) {
          const mid = (a + b) / 2;
          const yMid = this.evaluateAt(compiled, mid, angleMode);
          if (Math.abs(yMid) < 1e-12 || (b - a) / 2 < 1e-12) {
            root = mid;
            break;
          }
          const yA = this.evaluateAt(compiled, a, angleMode);
          if (yA * yMid < 0) {
            b = mid;
          } else {
            a = mid;
          }
        }
        if (root !== null) {
          const dist = Math.abs(root - guessX);
          if (dist < closestDist) {
            closestRoot = root;
            closestDist = dist;
          }
        }
      }
    }

    return closestRoot;
  }

  /**
   * Finds a local minimum or maximum in [xMin, xMax].
   * @param {string} type - 'min' or 'max'
   */
  static findExtremum(compiled, xMin, xMax, angleMode, type, guessX = 0) {
    const steps = 500;
    const dx = (xMax - xMin) / steps;
    const candidates = [];

    let prevY = null;
    let prevSlope = null;

    for (let i = 0; i <= steps; i++) {
      const x = xMin + i * dx;
      let y;
      try {
        y = this.evaluateAt(compiled, x, angleMode);
      } catch {
        continue;
      }
      if (isNaN(y) || !isFinite(y)) {
        prevY = null;
        prevSlope = null;
        continue;
      }

      if (prevY !== null) {
        const slope = (y - prevY) / dx;
        if (prevSlope !== null) {
          if (prevSlope > 0 && slope < 0) {
            if (type === 'max') {
              candidates.push({ x: x - dx / 2, y: prevY });
            }
          } else if (prevSlope < 0 && slope > 0) {
            if (type === 'min') {
              candidates.push({ x: x - dx / 2, y: prevY });
            }
          }
        }
        prevSlope = slope;
      }
      prevY = y;
    }

    if (candidates.length === 0) {
      return null;
    }

    let bestCandidate = null;
    let minDiff = Infinity;
    for (const cand of candidates) {
      const diff = Math.abs(cand.x - guessX);
      if (diff < minDiff) {
        minDiff = diff;
        bestCandidate = cand;
      }
    }

    let x = bestCandidate.x;
    let step = dx / 2;
    for (let iter = 0; iter < 30; iter++) {
      const y = this.evaluateAt(compiled, x, angleMode);
      const yLeft = this.evaluateAt(compiled, x - step, angleMode);
      const yRight = this.evaluateAt(compiled, x + step, angleMode);

      if (type === 'max') {
        if (isFinite(yLeft) && isFinite(yRight)) {
          if (yLeft > y && yLeft > yRight) {
            x = x - step;
          } else if (yRight > y && yRight > yLeft) {
            x = x + step;
          } else {
            step /= 2;
          }
        } else {
          step /= 2;
        }
      } else {
        if (isFinite(yLeft) && isFinite(yRight)) {
          if (yLeft < y && yLeft < yRight) {
            x = x - step;
          } else if (yRight < y && yRight < yLeft) {
            x = x + step;
          } else {
            step /= 2;
          }
        } else {
          step /= 2;
        }
      }
    }

    const finalY = this.evaluateAt(compiled, x, angleMode);
    if (isNaN(finalY) || !isFinite(finalY)) return null;
    return { x, y: finalY };
  }

  /**
   * Calculates dy/dx at xVal
   */
  static derivative(compiled, xVal, angleMode) {
    const h = 1e-5;
    try {
      const y1 = this.evaluateAt(compiled, xVal - h, angleMode);
      const y2 = this.evaluateAt(compiled, xVal + h, angleMode);
      if (isNaN(y1) || isNaN(y2) || !isFinite(y1) || !isFinite(y2)) return NaN;
      return (y2 - y1) / (2 * h);
    } catch {
      return NaN;
    }
  }

  /**
   * Definite integration from lower to upper bound using Trapezoidal Rule
   */
  static integrate(compiled, lower, upper, angleMode) {
    const N = 1000;
    const h = (upper - lower) / N;
    try {
      const yStart = this.evaluateAt(compiled, lower, angleMode);
      const yEnd = this.evaluateAt(compiled, upper, angleMode);

      if (isNaN(yStart) || isNaN(yEnd) || !isFinite(yStart) || !isFinite(yEnd)) return NaN;

      let sum = 0.5 * (yStart + yEnd);
      for (let i = 1; i < N; i++) {
        const yVal = this.evaluateAt(compiled, lower + i * h, angleMode);
        if (isNaN(yVal) || !isFinite(yVal)) return NaN;
        sum += yVal;
      }
      return sum * h;
    } catch {
      return NaN;
    }
  }

  /**
   * Calculates the second derivative f''(x) using a central second-order finite difference
   */
  static secondDerivative(compiled, xVal, angleMode) {
    const h = 1e-3;
    try {
      const yPlus = this.evaluateAt(compiled, xVal + h, angleMode);
      const yMinus = this.evaluateAt(compiled, xVal - h, angleMode);
      const yVal = this.evaluateAt(compiled, xVal, angleMode);
      if (isNaN(yPlus) || isNaN(yMinus) || isNaN(yVal) || !isFinite(yPlus) || !isFinite(yMinus) || !isFinite(yVal)) return NaN;
      return (yPlus - 2 * yVal + yMinus) / (h * h);
    } catch {
      return NaN;
    }
  }

  /**
   * Finds the exact inflection point in [x1, x2] using bisection on the second derivative
   */
  static findInflectionPoint(compiled, x1, x2, angleMode) {
    let a = x1;
    let b = x2;
    let lastMid = (a + b) / 2;
    for (let iter = 0; iter < 15; iter++) {
      const mid = (a + b) / 2;
      const fDoubleMid = this.secondDerivative(compiled, mid, angleMode);
      const fDoubleA = this.secondDerivative(compiled, a, angleMode);
      if (isNaN(fDoubleMid) || isNaN(fDoubleA) || !isFinite(fDoubleMid) || !isFinite(fDoubleA)) {
        return lastMid;
      }
      if (Math.abs(fDoubleMid) < 1e-8) {
        return mid;
      }
      if (fDoubleA * fDoubleMid < 0) {
        b = mid;
      } else {
        a = mid;
      }
      lastMid = mid;
    }
    return lastMid;
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = CalcEngine;
} else {
  window.CalcEngine = CalcEngine;
}
